src/libraries/Microsoft.Extensions.Caching.Memory/tests/MemoryCacheSetAndRemoveTests.cs

```diff
@@ -543,7 +543,6 @@ public void SetGetAndRemoveWorksWithObjectKeysWhenDifferentReferences()
         }
 
         [Fact]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/33993")]
```

Comment:
What changed in this test to allow it to be re-enabled?

Reply:
that was the test I was referring to above

> Note that at least one test disabled against https://github.com/dotnet/runtime/issues/33993 did not actually read the size. It would have failed due to some unrelated reason - back in 2020. I'm going to assume it was fixed by another change such as possibly https://github.com/dotnet/runtime/pull/42355.

it passes for me, so not sure what else I can do.

Reply:
Have you seen https://github.com/dotnet/runtime/issues/33993?

```
Assert.False() Failure
Expected: False
Actual:   True

at Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetAndSet_AreThreadSafe_AndUpdatesNeverLeavesNullValues() in /_/src/libraries/Microsoft.Extensions.Caching.Memory/tests/MemoryCacheSetAndRemoveTests.cs:line 588
```

Reply:
Thanks for the pointer. I'll re-disable but against a new issue, for clarity.

Reply:
Now, that's a pure speculation, but from my understanding, what's what might be happening:

1. Thread 1 sets a new entry. It happened to find a previous entry under the same key, so it attempts to expire it.

https://github.com/dotnet/runtime/blob/cea17b2ac2886a154aa813bcfa2fac386ed9f10a/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L129-L132

2. While expiring the previous entry, jit is (I think) allowed to rewrite it from

https://github.com/dotnet/runtime/blob/cea17b2ac2886a154aa813bcfa2fac386ed9f10a/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L237-L241

to

```csharp
var currentEvictionReason = EvictionReason;
_isExpired = true; 
if (currentEvictionReason == EvictionReason.None)
{
    EvictionReason = reason;
}
```

3. Thread 2 attempts to get an entry from the cache.

https://github.com/dotnet/runtime/blob/cea17b2ac2886a154aa813bcfa2fac386ed9f10a/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L213-L218

While doing so, it first check whether the entry is expired, and if so, the reason for the expiration.

4. Now we have all the parts. If thread 1 did manage to set `_isExpired` flag to `true`, but not yet `EvictionReason`, then thread 2 might be able to get `true` from `CheckExpired`, and read `EvictionReason.None` from `EvictionReason`. And this will lead `result` being `null`.

If so, we can test this theory by adding a `MemoryBarrier` between `EvictionReason` and `_isExpired`, like this:

```csharp
if (EvictionReason == EvictionReason.None)
{
	EvictionReason = reason;
}
Interlocked.MemoryBarrier();
_isExpired = true;
```

Reply:
That is plausible @vonzshik thank you. I'll link from the issue..